cpp/src/lance/io/scanner.cc

```diff
@@ -22,45 +22,59 @@
 
 #include <algorithm>
 #include <future>
-#include <set>
 #include <tuple>
 
 #include "lance/format/metadata.h"
 #include "lance/format/schema.h"
 #include "lance/io/filter.h"
+#include "lance/io/limit.h"
 #include "lance/io/project.h"
 #include "lance/io/reader.h"
 
 namespace lance::io {
 
 Scanner::Scanner(std::shared_ptr<FileReader> reader,
-                 std::shared_ptr<arrow::dataset::ScanOptions> options) noexcept
-    : reader_(reader), options_(options) {}
+                 std::shared_ptr<arrow::dataset::ScanOptions> options,
+                 std::optional<int64_t> limit,
+                 int64_t offset) noexcept
+    : reader_(reader),
+      options_(options),
+      limit_(limit),
+      offset_(offset),
+      max_queue_size_(static_cast<std::size_t>(options->batch_readahead)) {}
 
 Scanner::Scanner(const Scanner& other) noexcept
     : reader_(other.reader_),
       options_(other.options_),
+      limit_(other.limit_),
+      offset_(other.offset_),
       schema_(other.schema_),
       project_(other.project_),
-      current_chunk_(other.current_chunk_) {}
+      current_chunk_(other.current_chunk_),
+      max_queue_size_(other.max_queue_size_) {}
 
 Scanner::Scanner(Scanner&& other) noexcept
     : reader_(std::move(other.reader_)),
       options_(std::move(other.options_)),
+      limit_(other.limit_),
+      offset_(other.offset_),
       schema_(std::move(other.schema_)),
       project_(std::move(other.project_)),
       current_chunk_(other.current_chunk_),
+      max_queue_size_(other.max_queue_size_),
       q_(std::move(other.q_)) {}
 
 ::arrow::Status Scanner::Open() {
   schema_ = std::make_shared<lance::format::Schema>(reader_->schema());
-  ARROW_ASSIGN_OR_RAISE(project_, Project::Make(schema_, options_));
+  ARROW_ASSIGN_OR_RAISE(project_, Project::Make(schema_, options_, limit_, offset_));
+  if (!project_->CanParallelScan()) {
```

Comment:
We need to implement some data structure that tracks some mapping of chunk no -> offset/length? So there will be some "over reading" that gets discarded?

Reply:
Could you elaborate a bit? I am confused. In what case will `over-reading` occur?

Reply:
Parallel

Reply:
I see. Did you mean hedged read?  Yes, we can do hedged read later, we can discuss the design choices in https://github.com/eto-ai/lance/issues/48.